src/main/java/fi/aalto/cs/apluscourses/utils/CoursesClient.java

```diff
@@ -289,4 +292,30 @@ private static void requireResponseEntity(@NotNull HttpResponse response)
       throw new UnexpectedResponseException(response, "Response is missing body");
     }
   }
+
+  @NotNull
+  private static CloseableHttpClient createClient() {
+    return HttpClients.custom().setUserAgent(getUserAgent()).build();
+  }
+
+  @NotNull
+  private static String getUserAgent() {
+    ApplicationInfo appInfo = ApplicationInfo.getInstance();
+    ApplicationNamesInfo appNamesInfo = ApplicationNamesInfo.getInstance();
+    String intellijVersion = appInfo.getFullVersion();
+    String product = appNamesInfo.getProductName(); // E.g. "IDEA"
+    String edition = appNamesInfo.getEditionName(); // E.g. "Community Edition"
+    String os = System.getProperty("os.name");
+    String pluginVersion = BuildInfo.INSTANCE.version.toString();
+    String apacheUserAgent =
+        VersionInfo.getUserAgent("Apache-HttpClient", "org.apache.http.client", VersionInfo.class);
+    return String.format(
+        "intellij/%s (%s; %s; %s) apluscourses/%s %s",
+        intellijVersion,
+        product,
+        edition,
+        os,
+        pluginVersion,
+        apacheUserAgent);
```

Comment:
I assume this is all fail-safe if any of those return null?

Reply:
Also just a minor thing: all of this information is now fetched for every request, but it shouldn't change during the runtime of the application

Reply:
`String.format("Hello %s!", null)` returns `"Hello null!", so it would leave "null"s to the user agent but not fail.

Yes, the whole user agent string could be computed only once (perhaps in a lazy way?), unless we add course information there (which would need also more changes). If we are sure we are not gonna need course info in user-agent, I'll make it a static lazy-initialized string.

Reply:
We can switch it again if want to add course info, so let's go with static lazy-initialized for now then